docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,827 @@
+Thin-edge doesn't enforce/dictate how a user uses these namespace levels.
+But there are conventions that can be followed to cover a wide range of use-cases, as follows:
+
+* `te/<device-namespace>/<device-id>/<service-namespace>/<service-id>/...`
+* `te/<device-namespace>/<device-id>/<device-local-namespace>/<device-local-component-id>/...`
+* `te/<entity-namespace>/<entity-id>/<component-namespace>/<component-id>/...`
+* `te/<device-family>/<device-series>/<part-series>/<part-id>/...`
+
+Even for the data type levels, a user is free to define those as they wish.
+But thin-edge has some pre-defined `<data-type>` subtopics for well-known types like:
+`measurements`, `alarms`, `events` and `commands`, on which it enforces some constraints.
+
+**Subtopics for predefined telemetry data**
+
+|Type|Example|
+|------|-------|
+|Measurements|`<entity_id_prefix>/mes/[/<measurement-type>]`|
+|Events|`<entity_id_prefix>/evn/<event-type>`|
+|Alarms|`<entity_id_prefix>/alr/<alarm-type>`|
+
+**Subtopics for predefined commands**
+
+|Type|Example|
+|------|-------|
+|Software List|`<entity_id_prefix>/cmd/software_list`|
+|Software Update|`<entity_id_prefix>/cmd/software_update`|
+|Configuration Snapshot|`<entity_id_prefix>/cmd/config_snapshot`|
+|Configuration Update|`<entity_id_prefix>/cmd/config_update`|
+|Firmware Update|`<entity_id_prefix>/cmd/firmware_update`|
+|Restart[^1]|`<entity_id_prefix>/cmd/restart`|
+
+[^1]: Restart would mean a device restart or a service restart based on the target entity
+
+### Entity registration
+
+Since thin-edge doesn't enforce what each device identification level means,
+an explicit registration is required to register every entity that is going to send data or receive commands.
+For example, before a measurement can be sent from a service named `tedge-agent` from the device `Rpi1001`,
+the entity named `Rpi1001` must be registered as a `device`, and `tedge-agent` must be registered as a `service`.
+
+An entity can be registered with thin-edge by publishing a retained message to the entity identification topic prefix
+with the entity type and other metadata that defines that entity.
+To model the example mentioned above, if an entity identification topic scheme like the following is used:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+The `Rpi1001` device must be registered as a `device` by publishing the following retained message:
+
+```sh te2mqtt
+tedge mqtt pub -r te/_/Rpi1001 '{
+  "type": "device"
+}'
+```
+
+The device is registered under the default namespace indicated by `_` at the second level.
+The registration message supports additional fields like `name`, `parent`, `device-type` etc as well,
+which will be explained in detail later.
+
+Once the device is registered, the `tedge-agent` service can be registered as a `service` type as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/_/Rpi1001/_/tedge-agent '{
+  "type": "service"
+}'
+```
```

Comment:
There is something implied here: the service is attached to the device `te/_/Rpi1001` because the later has already been registered.

I'm okay with that but this need to be said in this document:
- Do we provide two ways to declare that P is a parent of C?
- Using topic prefix (as here)?
- Using some metadata property?